src/Compilers/CSharp/Test/Symbol/Symbols/ModuleInitializers/ModuleInitializersTests.cs

```diff
@@ -778,5 +778,169 @@ namespace System.Runtime.CompilerServices { class ModuleInitializerAttribute : S
+
+            var s2 = @"
+using System;
+using System.Runtime.CompilerServices;
+
+public class B
+{
+    [ModuleInitializer]
+    public static void M2()
+    {
+        Console.Write(2);
+    }
+}";
+            var comp2 = CreateCompilation(s2, options: moduleOptions.WithModuleName("B"), parseOptions: s_parseOptions, references: new[] { ref1 });
+            comp2.VerifyDiagnostics();
+            var ref2 = comp2.EmitToImageReference();
+            CompileAndVerify(comp2, symbolValidator: validateModuleInitializer, verify: Verification.Skipped);
+
+            var exeOptions = TestOptions.ReleaseExe
+                .WithMetadataImportOptions(MetadataImportOptions.All)
+                .WithModuleName("C");
+            var s3 = @"
+using System;
+
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        Console.Write(3);
+    }
+}";
+            var comp3 = CreateCompilation(s3, options: exeOptions, parseOptions: s_parseOptions, references: new[] { ref1, ref2 });
+            comp3.VerifyDiagnostics();
+            CompileAndVerify(comp3, symbolValidator: validateNoModuleInitializer, expectedOutput: "3");
+
+            var s4 = @"
+using System;
+
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        new A();
+        new B();
+        Console.Write(3);
+    }
+}";
+            var comp4 = CreateCompilation(s4, options: exeOptions, parseOptions: s_parseOptions, references: new[] { ref1, ref2 });
+            comp4.VerifyDiagnostics();
+            CompileAndVerify(comp4, symbolValidator: validateNoModuleInitializer, expectedOutput: "123");
+
+            var s5 = @"
+using System;
+
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        new B();
+        Console.Write(3);
+        new A();
+    }
+}";
+            var comp5 = CreateCompilation(s5, options: exeOptions, parseOptions: s_parseOptions, references: new[] { ref1, ref2 });
+            comp5.VerifyDiagnostics();
+            // This order seems surprising, but is likely related to the order in which types are loaded when a method is called.
+            CompileAndVerify(comp5, symbolValidator: validateNoModuleInitializer, expectedOutput: "213");
+
+            var s6 = @"
+using System;
+
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        new A();
+        Console.Write(3);
+    }
+}";
+            var comp6 = CreateCompilation(s6, options: exeOptions, parseOptions: s_parseOptions, references: new[] { ref1, ref2 });
+            comp6.VerifyDiagnostics();
+            CompileAndVerify(comp6, symbolValidator: validateNoModuleInitializer, expectedOutput: "13");
+
+            var s7 = @"
+using System;
+using System.Runtime.CompilerServices;
+
+public class Program
+{
+    [ModuleInitializer]
+    public static void Init()
+    {
+        Console.Write(1);
```

Comment:
>1 [](start = 22, length = 1)

Does this number conflict with the pone printed by ```ref1```? Please use unique numbers to avoid any confusion.